README.md
# Zinc64

[![Build Status](https://travis-ci.org/digitalstreamio/zinc64.svg?branch=master)](https://travis-ci.org/digitalstreamio/zinc64)
[![Crates.io](https://img.shields.io/crates/v/zinc64.svg?maxAge=2592000)](https://crates.io/crates/zinc64)

** **NOTE: zinc64 crate has been renamed to zinc64-emu** **

## Overview

zinc64 is a Commodore 64 emulator toolkit "with batteries included but
swappable". It is designed to be used as a standalone emulator or a library
used to build new emulators. The design philosophy allows for each component
to be swapped out and replaced by different implementation. Therefore,
special considerations were made to model interactions between chips
without coupling them together.

It implements MOS 6510 CPU, MOS 6526 CIA, MOS 6581 SID,
MOS 6567/6569 VIC chipset as well as various devices and peripherals available
with C64.

### Story

zinc64 was started as an exercise to learn Rust and explore Commodore 64
hardware in more detail. Somewhere around mid 2016 I needed to feed my 8-bit
nostalgia so I picked up a working Commodore 64 (physical version) and started
to assemble various accessories required to get software onto it. Soon enough I
had picked up a copy of C64 Programmer's Reference Guide and the rest is now
history.

2019 is bringing support for revised SDL port with console support and
bare-metal environments, more specifically a bare-metal Raspberry Pi 3 port.
See zinc64-rpi for an early preview.

## Design

### Feature `std`

zinc64-emu crate works without the standard library, such as in bare-metal environments. To use zinc64-emu in a #[no_std] environment, use: 

```toml
[dependencies]
zinc64-emu = { version = "0.8.0", default-features = false }
```

### Extensibility

The emulator components may be swapped out by providing custom core::ChipFactory trait
implementation. The default implementation of the core::ChipFactory trait is done through
system::C64Factory. The chip factory object is passed into system::C64 component
that provides core emulator functionality.

Here is an example how these components are used together:

        let config = Rc::new(Config::new(SystemModel::from("pal")));
        let chip_factory = Box::new(C64Factory::new(config.clone()));
        let mut c64 = C64::new(config.clone(), chip_factory).unwrap();
        c64.reset(true);

The four core traits used to model system operation are Chip, Cpu, Mmu and Addressable.

    /// A chip represents a system component that is driven by clock signal.
    pub trait Chip {
        /// The core method of the chip, emulates one clock cycle of the chip.
        fn clock(&mut self);
        /// Process delta cycles at once.
        fn clock_delta(&mut self, delta: u32);
        /// Handle vsync event.
        fn process_vsync(&mut self);
        /// Handle reset signal.
        fn reset(&mut self);
        // I/O
        /// Read value from the specified register.
        fn read(&mut self, reg: u8) -> u8;
        /// Write value to the specified register.
        fn write(&mut self, reg: u8, value: u8);
    }

    /// CPU is responsible for decoding and executing instructions.
    pub trait Cpu {
        ...
        /// The core method of the cpu, decodes and executes one instruction. Tick callback is invoked
        /// for each elapsed clock cycle.
        fn step(&mut self, tick_fn: &TickFn);
        // I/O
        /// Read byte from the specified address.
        fn read(&self, address: u16) -> u8;
        /// Write byte to the specified address.
        fn write(&mut self, address: u16, value: u8);
    }

    /// Represents memory management unit which controls visible memory banks
    /// and is used by CPU to read from and write to memory locations.
    pub trait Mmu {
        /// Change bank configuration based on the specified mode.
        fn switch_banks(&mut self, mode: u8);
        // I/O
        /// Read byte from the specified address.
        fn read(&self, address: u16) -> u8;
        /// Write byte to the specified address.
        fn write(&mut self, address: u16, value: u8);
    }

    /// Addressable represents a bank of memory.
    pub trait Addressable {
        /// Read byte from the specified address.
        fn read(&self, address: u16) -> u8;
        /// Write byte to the specified address.
        fn write(&mut self, address: u16, value: u8);
    }

Since all system components with the exception of Cpu and Mmu implement Chip trait,
interactions between chips and other components are limited to and handled through
shared I/O lines/pins that are provided to chip constructors. This allows implementation
of chips to be decoupled from each other.

## Status

| Class    | Component     | Status      |
|----------|---------------|-------------|
| Chipset  | 6510 CPU      | Done
| Chipset  | Memory        | Done
| Chipset  | 6526 CIA      | Done
| Chipset  | 6581 SID      | Done
| Chipset  | 6567 VIC      | Done
| Device   | Cartridge     | Done
| Device   | Floppy        | Not Started
| Device   | Datassette    | Done
| Device   | Keyboard      | Done
| Device   | Joystick      | Done
| Device   | Mouse         | Not Started
| Debugger | Remote        | Done
| Debugger | Radare2       | Done
| Format   | Bin           | Done
| Format   | Crt           | Done
| Format   | D64           | Not Started
| Format   | P00           | Done
| Format   | Prg           | Done
| Format   | Tap           | Done
| Format   | T64           | Not Started
| Client   | SDL           | In Progress
| Client   | Raspi3        | In Progress

## Roadmap

- v0.9  - rpi port
- v0.10  - floppy support

## Getting Started

1. Install Rust compiler or follow steps @ https://www.rust-lang.org/en-US/install.html.

        curl https://sh.rustup.rs -sSf | sh

2. Clone this repository.

        git clone https://github.com/digitalstreamio/zinc64

	or download as zip archive

		https://github.com/digitalstreamio/zinc64/archive/master.zip

3. Build the emulator.

        cd zinc64
        cargo build --release --all

4. Run the emulator.

        ./target/release/zinc64

    or start a program

    	./target/release/zinc64 --autostart path

### Windows Considerations

1. Install [Microsoft Visual C++ Build Tools 2017](https://www.visualstudio.com/downloads/#build-tools-for-visual-studio-2017). Select Visual C++ build tools workload.

2. Install [SDL2 Development Libraries](http://www.libsdl.org/release/SDL2-devel-2.0.7-VC.zip).

3. Copy all SDL2 lib files from

		SDL2-devel-2.0.x-VC\SDL2-2.0.x\lib\x64\

	to 

		C:\Users\{Your Username}\.rustup\toolchains\stable-x86_64-pc-windows-msvc\lib\rustlib\x86_64-pc-windows-msvc\lib

4. Copy SDL2.dll from

		SDL2-devel-2.0.x-VC\SDL2-2.0.x\lib\x64\

	to zinc64 project directory

## Debugger

To start the debugger, run the emulator with '-d' or '--debug' option. Optionally, you can specify '--debugaddress'
to bind to a specific address.

        ./target/release/zinc64 --debug

To connect to the debugger, telnet to the address and port used by the debugger.

        telnet localhost 9999

Debugger commands and syntax are modeled after Vice emulator.  To see a list of available commands,
type in the debugging session:

        help

or to get help on a specific command:

        help <command>

### Radare2

Initial support for radare2 has been merged in version 0.3. To start the emulator with RAP server support, run

        ./target/release/zinc64 --rap 127.0.0.1:9999

and connect with

        radare2 -a 6502 -d rap://localhost:9999/1

## Examples

I've included a number of examples from Kick Assembler that I've used to test various components of the emulator. They can be found in the bin folder of this repository and started with the emulator's autostart option.

        ./target/release/zinc64 --autostart bin/SineAndGraphics.prg

| Program                  | Status  |
|--------------------------|---------|
| 6502_functional_test.bin | Pass    |
| FloydSteinberg.prg       | Pass    | 
| KoalaShower.prg          | Pass    |
| Message.prg              | Pass    |
| MusicIrq.prg             | Pass    |
| Scroll.prg               | Pass    |
| SID_Player.prg           | Pass    |
| SimpleSplits.prg         | Fails   |
| SineAndGraphics.prg      | Pass    |

### Tests

The cpu validation was performed with the help of [Klaus2m5 functional tests](https://github.com/Klaus2m5/6502_65C02_functional_tests) for the 6502 processor 

        ./target/release/zinc64 --binary bin/6502_functional_test.bin --offset=1024 --console --loglevel trace

## Keyboard Shortcuts

| Shortcut  | Function          |
|-----------|-------------------|
| Escape    | Console
| Alt-Enter | Toggle Full Screen
| Alt-F9    | Reset
| Alt-H     | Activate Debugger
| Alt-M     | Toggle Mute
| Alt-P     | Toggle Pause
| Alt-Q     | Quit
| Alt-W     | Warp Mode
| Ctrl-F1   | Tape Play/Stop
| NumPad-2  | Joystick Bottom
| NumPad-4  | Joystick Left
| NumPad-5  | Joystick Fire
| NumPad-6  | Joystick Right
| NumPad-8  | Joystick Top

## Credits

- Commodore folks for building an iconic 8-bit machine
- Rust developers for providing an incredible language to develop in
- Thanks to Rafal Wiosna for passing onto me some of his passion for 8-bit machines ;)
- Thanks to Klaus Dormann for his 6502_65C02_functional_tests, without which I would be lost
- Thanks to Dag Lem for his reSID implementation
- Thanks to Christian Bauer for his wonderful "The MOS 6567/6569 video controller (VIC-II) and its application in the Commodore 64" paper
- Thanks to Peter Schepers for his "Introduction to the various Emulator File Formats"
- Thanks to c64-wiki.com for my to go reference on various hardware components



CHANGELOG
# CHANGELOG

### Week 1 - 20161224

- landed cpu
- landed mem with basic bank switching
- landed skeleton c64
- milestone 1: pass 6502 functional test

### Week 2 - 20161231

- landed cia
- landed mem device i/o
- landed keyboard
- landed cli parsing
- landed preliminary loaders
- landed skeleton ui and event handling
- milestone 2: run keyboard_wait_s.hex
- landed vic device i/o
- landed pal config
- landed color ram and vic memory map
- landed render target and linked it to ui rendering
- landed vic text mode and vsync
- fix ...
- milestone 3: boot to basic

### Week 3 - 20170107

- landed ui options
- landed expansion port and cartridge
- revamped memory map
- landed joystick
- revamped cia port handling
- revamped loader support
- landed autostart
* hack up vic sprites
- milestone 4: run frogger_ii.crt

### Week 4 - 20170114

- landed logging
- landed app speed control
- landed cpu jam
- rework cia interrupt and step logic
- rework c64 frame execution logic
- rework c64 reset logic
- revamp cpu instruction tracing
- revamp cpu interrupt tracking, fixes frogger freeze
- landed cia tod
- landed datassette and tap loader
- rework vic drawing and add scrolling support
- milestone 5: run commando.tap

### Week 5 - 20170121

- ported resid
- landed sid
- milestone 6: run musicirq.prg

... LONG INTERMISSION ...

### Week 6 - 20171216

- landed compliant sid output

### Week 7 - 20171223

- landed revised cpu clock emulation
- landed revised keyboard handling

### Week 8 - 20171230

- landed revised model and system configuration
- landed revamped cli handling
- factored out app ui into separate components
- factored out common hw primitives into core module

### Week 9 - 20180106

- landed modularized emulator design

### Week 10 - 20180113

- fixed sound playback artifacts, hello pure sid playback
- milestone 7: run musicirq.prg without static/noise
- milestone 8: publish zinc64 crate

### Week 11 - 20180120

- landed revised cia clock emulation
- miletone 9: run cia1tab.p00

### Week 12 - 20180127

- landed debugger support

### Week 13 - 20180203

- landed nicer debugging support

### Week 14 - 20180210

- landed vic rewrite part 1
- milestone 10: run sid_player.prg

### Week 15 - 20180714

- landed vic border/graphics/sprite/mux units
- milestone 11: run superball.prg

### Week 16 - 20180721

- refactoring and cleanup

### Week 17 - 20180811

- ci and deployment for linux/windows

### Week 18 - 20181111

- migration to rust 2018

### Week 19 - 20190106

- add no_std build support

### Week 20 - 20190113

- add initial raspi3 build

### Week 21 - 20190120

- add raspi video output

### Week 22 - 20190127

- add raspi loader support

### Week 23 - 20190210

- add raspi fat32 support

## Week 24 - 20190303

- add raspi sound output

## Week 25 - 20190324

- landed console part 1

## Week 26 - 20190421

- landed console part 2

## Week 27 - 20190505

- landed cartridge mapping rewrite
- landed various game fixes
  - fixed cart memory mapping modes (sotb)
  - fixed sid reg reading (pang)
  - fixed kb scanning (rr)
  - fixed kb run/stop mapping (pitstop2)
  - fixed joy2 regression (sotb)
  - fixed cpu nmi handling (super mario)
  - fixed cpu opcode ane (1942)
- run sotb.crt



src/config.rs
/*
 * Copyright (c) 2016-2017 Sebastian Jastrzebski. All rights reserved.
 *
 * This file is part of zinc64.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

use device::joystick;
use util::{Dimension, Rect};

// http://codebase64.org/doku.php?id=base:cpu_clocking
// http://dustlayer.com/vic-ii/2013/4/25/vic-ii-for-beginners-beyond-the-screen-rasters-cycle
// http://www.antimon.org/dl/c64/code/missing.txt

/*
All clock frequencies in the C64 are derived from a single clock quartz which has the frequency of 4
times the frequency of the color carrier used for PAL or NTSC.

PAL C64 master clock: 17.734475 MHz

NTSC C64 master clock: 14.31818 MHz

The CPU frequency is then calculated from that by simply dividing the frequency by 18 (PAL) or 14
(NTSC). The VIC-II runs at a frequency which is exactly 8 times that of the CPU. This is the so called
“dot clock” which has to be very precise in order to keep the right timing needed to generate a video
signal compatible with all TVs. The CPU of the time could not go that fast, max. 1MHz, but the CPU
still needs to be phase synchronous to the VIC-II because they share control of the address/data bus
of the machine. That's why the VIC-II internally provides a clock divider which feeds the CPU.
*/

const CLOCK_MASTER_PAL: u32 = 17734475;
const CLOCK_CPU_PAL: u32 = CLOCK_MASTER_PAL / 18; // 985248 Hz
#[allow(dead_code)]
const CLOCK_VIC_PAL: u32 = CLOCK_CPU_PAL * 8; // 7881984 Hz

/*

The dimensions of the video display for the different VIC types are as
follows:

          | Video  | # of  | Visible | Cycles/ |  Visible
   Type   | system | lines |  lines  |  line   | pixels/line
 ---------+--------+-------+---------+---------+------------
 6567R56A | NTSC-M |  262  |   234   |   64    |    411
  6567R8  | NTSC-M |  263  |   235   |   65    |    418
   6569   |  PAL-B |  312  |   284   |   63    |    403

          | First  |  Last  |              |   First    |   Last
          | vblank | vblank | First X coo. |  visible   |  visible
   Type   |  line  |  line  |  of a line   |   X coo.   |   X coo.
 ---------+--------+--------+--------------+------------+-----------
 6567R56A |   13   |   40   |  412 ($19c)  | 488 ($1e8) | 388 ($184)
  6567R8  |   13   |   40   |  412 ($19c)  | 489 ($1e9) | 396 ($18c)
   6569   |  300   |   15   |  404 ($194)  | 480 ($1e0) | 380 ($17c)

*/

const DISPLAY_WIDTH_PAL: u16 = 504;
const DISPLAY_HEIGHT_PAL: u16 = 312;

const VISIBLE_WIDTH_PAL: u16 = 403;
const VISIBLE_HEIGHT_PAL: u16 = 284;
const VISIBLE_FIRST_COL_PAL: u16 = 80; // translated 76, original 480, but we offset x to beg of display
const VISIBLE_FIRST_LINE_PAL: u16 = 16;
#[allow(dead_code)]
const VISIBLE_LAST_COL_PAL: u16 = 484; // translated 480, original 380, but we offset x to beg of display
#[allow(dead_code)]
const VISIBLE_LAST_LINE_PAL: u16 = 299;

/*

The height and width of the display window can each be set to two different
values with the bits RSEL and CSEL in the registers $d011 and $d016:

 RSEL|  Display window height   | First line  | Last line
 ----+--------------------------+-------------+----------
   0 | 24 text lines/192 pixels |   55 ($37)  | 246 ($f6)
   1 | 25 text lines/200 pixels |   51 ($33)  | 250 ($fa)

 CSEL|   Display window width   | First X coo. | Last X coo.
 ----+--------------------------+--------------+------------
   0 | 38 characters/304 pixels |   31 ($1f)   |  334 ($14e)
   1 | 40 characters/320 pixels |   24 ($18)   |  343 ($157)

The X coordinates run up to $1ff (only $1f7 on the 6569) within a line, then comes X coordinate 0.

There are 2×2 comparators belonging to each of the two flip flops. There
comparators compare the X/Y position of the raster beam with one of two
hardwired values (depending on the state of the CSEL/RSEL bits) to control
the flip flops. The comparisons only match if the values are reached
precisely. There is no comparison with an interval.

The horizontal comparison values:

       |   CSEL=0   |   CSEL=1
 ------+------------+-----------
 Left  |  31 ($1f)  |  24 ($18)
 Right | 335 ($14f) | 344 ($158)

And the vertical ones:

        |   RSEL=0  |  RSEL=1
 -------+-----------+----------
 Top    |  55 ($37) |  51 ($33)
 Bottom | 247 ($f7) | 251 ($fb)

*/

const WINDOW_WIDTH: u16 = 320;
const WINDOW_HEIGHT: u16 = 200;
const WINDOW_FIRST_COL: u16 = 128; // translated 124, original 24
const WINDOW_FIRST_LINE: u16 = 51;
#[allow(dead_code)]
const WINDOW_LAST_LINE: u16 = 250;
#[allow(dead_code)]
const WINDOW_LAST_COL: u16 = 447; // translated 443, original 343

const RASTER_TIME_BYTE_PAL: u16 = 1; // 8 pixels/cpu cycle
const RASTER_LINE_CYCLES_PAL: u16 = DISPLAY_WIDTH_PAL / 8 * RASTER_TIME_BYTE_PAL; // 63
const RASTER_FRAME_CYCLES_PAL: u16 = DISPLAY_HEIGHT_PAL * RASTER_LINE_CYCLES_PAL; // 19656
const RASTER_REFRESH_RATE_PAL: f64 = (CLOCK_CPU_PAL as f64) / (RASTER_FRAME_CYCLES_PAL as f64); // 50.125

#[derive(Copy, Clone)]
pub struct Config {
    // Cpu
    pub cpu_frequency: u32,
    // Video
    pub raster_size: Dimension,
    pub screen_size: Dimension,
    pub screen: Rect,
    pub window_size: Dimension,
    pub graphics: Rect,
    pub window: Rect,
    pub frame_cycles: u32,
    pub frame_duration_ns: u32,
    pub raster_line_cycles: u16,
    pub refresh_rate: f64,
    // Devices
    pub joystick1: joystick::Mode,
    pub joystick2: joystick::Mode,
}

impl Config {
    pub fn new(model: &str) -> Config {
        match model {
            "pal" => Config::pal(),
            _ => panic!("invalid model {}", model),
        }
    }

    pub fn pal() -> Config {
        let display_size = Dimension::new(DISPLAY_WIDTH_PAL, DISPLAY_HEIGHT_PAL);
        let visible_size = Dimension::new(VISIBLE_WIDTH_PAL, VISIBLE_HEIGHT_PAL);
        let window_size = Dimension::new(WINDOW_WIDTH, WINDOW_HEIGHT);
        let graphics = Rect::new_with_dim(WINDOW_FIRST_COL, WINDOW_FIRST_LINE - 3, window_size);
        let window = Rect::new_with_dim(
            WINDOW_FIRST_COL - VISIBLE_FIRST_COL_PAL,
            WINDOW_FIRST_LINE - VISIBLE_FIRST_LINE_PAL,
            window_size,
        );
        Config {
            cpu_frequency: CLOCK_CPU_PAL,
            raster_size: display_size,
            screen_size: visible_size,
            screen: Rect::new_with_dim(VISIBLE_FIRST_COL_PAL, VISIBLE_FIRST_LINE_PAL, visible_size),
            graphics: graphics,
            window_size: window_size,
            window: window,
            frame_cycles: RASTER_FRAME_CYCLES_PAL as u32,
            frame_duration_ns: ((1.0 / RASTER_REFRESH_RATE_PAL) * 1_000_000_000.0) as u32,
            raster_line_cycles: RASTER_LINE_CYCLES_PAL,
            refresh_rate: RASTER_REFRESH_RATE_PAL,
            joystick1: joystick::Mode::Numpad,
            joystick2: joystick::Mode::None,
        }
    }
}
